File: php/testcode/xxe_034.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_xxe_json_only_api
function xxe034(BenchmarkRequest $req): BenchmarkResponse {
    $ct = $req->header('Content-Type');
    if (strpos($ct, 'application/json') === false) {
        return BenchmarkResponse::badRequest('json only');
    }
    $data = json_decode($req->bodyStr(), true); // vuln-code-snippet safe-line php_xxe_json_only_api
    return BenchmarkResponse::ok(json_encode($data));
}
// vuln-code-snippet end php_xxe_json_only_api

File: ruby/testcode/xxe_034.rb
require_relative 'shared'
require 'rexml/document'
require 'rexml/sax2listener'
require 'rexml/parsers/sax2parser'

# vuln-code-snippet start ruby_xxe_sax2_rexml
def parse_sax2_rexml_handler(req)
  xml = req.post('xml')
  class MyListener; include REXML::SAX2Listener; end
  parser = REXML::SAX2Parser.new(xml) # vuln-code-snippet vuln-line ruby_xxe_sax2_rexml
  parser.parse
  BenchmarkResponse.ok('parsed')
end
# vuln-code-snippet end ruby_xxe_sax2_rexml
